auto-self-nominate/index.ts

```diff
@@ -0,0 +1,179 @@
+            return "bc";
+        case "corgi":
+            return "wc";
+        case "mainnet":
+            return "cc";
+        default:
+            throw Error("Invalid server configuration");
+    }
+}
+
+const SERVER: string = (() => {
+    const server = process.env.SERVER || "beagle";
+    if (["beagle", "corgi", "mainnet"].indexOf(server) >= 0) {
+        return server;
+    } else {
+        throw Error("Invalid server configuration");
+    }
+})();
+
+const sdk = (() => {
+    console.log(`sdk ${SERVER} ${process.env.RPC_URL}`);
+    return new SDK({
+        server: process.env.RPC_URL || rpcUrl(SERVER),
+        networkId: networkId(SERVER),
+    });
+})();
+
+function decodePlatformAddressFromPubkey(buffer: Buffer): PlatformAddress {
+    const pubkey = buffer.toString("hex");
+    return PlatformAddress.fromPublic(pubkey, {
+        networkId: sdk.networkId,
+    });
+}
+
+function decodeNumber(buffer: Buffer): number {
+    return parseInt("0x" + buffer.toString("hex"), 16);
+}
+
+async function getCurrentTermId(bestBlockNumber: number): Promise<number | null> {
+    return new Promise((resolve, reject) => {
+        sdk.rpc
+            .sendRpcRequest("chain_getTermMetadata", [bestBlockNumber])
+            .then(result => {
+                if (result === null) {
+                    resolve(null);
+                }
+                const [prevTermLastBlockNumber, currentTermId] = result;
+                if (
+                    typeof prevTermLastBlockNumber === "number" &&
+                    typeof currentTermId === "number"
+                ) {
+                    resolve(currentTermId);
+                }
+                reject(
+                    Error(
+                        `Expected getTermMetadata to return [number, number] | null but it returned ${result}`,
+                    ),
+                );
+            })
+            .catch(reject);
+    });
+}
+
+async function sendSelfNominateTransaction(
+    accountInfo: { account: string; passphrase: string },
+    params: { deposit: U64; metadata: string },
+) {
+    const tx = createSelfNomination(params);
+    const { account, passphrase } = accountInfo;
+    await sdk.rpc.account.sendTransaction({
+        tx,
+        account,
+        passphrase,
+    });
+}
+
+function createSelfNomination(params: { deposit: U64; metadata: string }): Custom {
+    const { deposit, metadata } = params;
+    const handlerId = new U64(STAKE_ACTION_HANDLER_ID);
+    const actionTag = ACTION_TAG_SELF_NOMINATE;
+    const bytes = RLP.encode([actionTag, deposit, metadata]);
+    return new Custom(
+        {
+            handlerId,
+            bytes,
+        },
+        networkId(SERVER),
+    );
+}
+
+async function needsNomination(
+    address: string,
+    nominationFrequencyLevel: number,
+): Promise<boolean> {
+    const bestBlockNumber = await sdk.rpc.chain.getBestBlockNumber()!;
+    const candidates: Map<string, number> = await (async () => {
```

Comment:
Usually, you don't need to use `Map`, if the type of key is `string` and you don't iterate over it.

Reply:
I fixed the type of candidates to `{ [key: string] : number}`